### PR TITLE
Have reduction observer write data from doubles and/or std vectors

### PR DESCRIPTION
**Pull request by carmaza**
## Proposed changes

This is needed for observing reductions of integrals/norms of tensors component-wise.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/IO/Observer/ReductionActions.hpp [hunk 1]
```diff
@@ -163,6 +163,19 @@ namespace ThreadedActions {
...
 struct WriteReductionData {
  private:
+  static void append_to_reduction_data(
+      gsl::not_null<std::vector<double>*> all_reduction_data, const double t) {
+    all_reduction_data->push_back(t);
+  }
+
+  static void append_to_reduction_data(
+      gsl::not_null<std::vector<double>*> all_reduction_data,
+      const std::vector<double>& t) {
```
Comment (nilsvu) [inline src/IO/Observer/ReductionActions.hpp]: Thanks for working on observing integrals so quickly! To make sure I understand: Are you planning to use vectors to represent the integral over each component of a tensor? If so, does sending vectors have an advantage over using a `ReductionDatum` with a `double` for each component?

Comment (carmaza) [reply src/IO/Observer/ReductionActions.hpp]: Each integral will be `push_back`'d to a `std::vector` containing all the integrals. I actually couldn't make the individual doubles work for tensors of rank > 0. Look at #2048 and let me know what you think.

File: src/IO/Observer/ReductionActions.hpp [hunk 2]
```diff
@@ -163,6 +163,19 @@ namespace ThreadedActions {
  */
 struct WriteReductionData {
  private:
+  static void append_to_reduction_data(
+      gsl::not_null<std::vector<double>*> all_reduction_data, const double t) {
```
Comment (nilsdeppe) [inline src/IO/Observer/ReductionActions.hpp]: `const gsl::not_null` and noexcept

File: src/IO/Observer/ReductionActions.hpp [hunk 3]
```diff
@@ -163,6 +163,19 @@ namespace ThreadedActions {
...
+    all_reduction_data->push_back(t);
+  }
+
+  static void append_to_reduction_data(
+      gsl::not_null<std::vector<double>*> all_reduction_data,
```
Comment (nilsdeppe) [inline src/IO/Observer/ReductionActions.hpp]: `const gsl::not_null` and noexcept

File: src/IO/Observer/ReductionActions.hpp [hunk 4]
```diff
@@ -163,6 +163,19 @@ namespace ThreadedActions {
...
+      const std::vector<double>& t) {
+    for (const double p : t) {
```
Comment (nilsdeppe) [inline src/IO/Observer/ReductionActions.hpp]: Probably `all_reduction_data->insert(all_reduction_data->-end(), t.begin(), t.end());` is the "modern" way of doing this

File: tests/Unit/IO/Observers/ObserverHelpers.hpp [hunk 5]
```diff
@@ -94,11 +94,37 @@ struct observer_writer_component {
...
     Parallel::ReductionDatum<double, funcl::AssertEqual<>>,
     Parallel::ReductionDatum<size_t, funcl::Plus<>>, l2_error_datum,
     l2_error_datum>;
 
+struct VectorPlus {
+  std::vector<double> operator()(const std::vector<double>& lhs,
+                                 const std::vector<double>& rhs) const
+      noexcept {
+    ASSERT(lhs.size() == rhs.size(),
+           "Error: vector sizes in `VectorPlus` operator do not match.");
```
Comment (nilsdeppe) [inline tests/Unit/IO/Observers/ObserverHelpers.hpp]: Don't need the `Error:`, that's printed as part of the `ASSERT`. Could you also output the size of `lhs` and `rhs`

File: tests/Unit/IO/Observers/Test_ReductionObserver.cpp [hunk 6]
```diff
@@ -83,73 +84,161 @@ SPECTRE_TEST_CASE("Unit.IO.Observers.ReductionObserver", "[Unit][Observers]") {
...
-    file_system::rm(h5_file_name, true);
-  }
+  const auto make_fake_reduction_data = make_overloader(
+      [](const observers::ArrayComponentId& id, const double time,
+         const helpers::reduction_data_from_doubles& /*meta*/) noexcept {
+        const auto hashed_id =
+            static_cast<double>(std::hash<observers::ArrayComponentId>{}(id));
+        constexpr size_t number_of_grid_points = 4;
+        const double error0 = 1.0e-10 * hashed_id + time;
+        const double error1 = 1.0e-12 * hashed_id + 2 * time;
```
Comment (nilsdeppe) [inline tests/Unit/IO/Observers/Test_ReductionObserver.cpp]: 2 -> 2.0 (and below as well)

File: tests/Unit/IO/Observers/Test_ReductionObserver.cpp [hunk 7]
```diff
@@ -83,73 +84,161 @@ SPECTRE_TEST_CASE("Unit.IO.Observers.ReductionObserver", "[Unit][Observers]") {
...
 
-  const double time = 3.;
-  const std::vector<std::string> legend{"Time", "NumberOfPoints", "Error0",
-                                        "Error1"};
-  // Test passing reduction data.
-  for (const auto& id : element_ids) {
-    const observers::ArrayComponentId array_id(
-        std::add_pointer_t<element_comp>{nullptr},
-        Parallel::ArrayIndex<ElementIndex<2>>{ElementIndex<2>{id}});
+    const double time = 3.;
```
Comment (nilsdeppe) [inline tests/Unit/IO/Observers/Test_ReductionObserver.cpp]: `3.` -> `3.0` for consistency

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
I'm happy for you to squash immediately

**Comment by carmaza** [comment]
Squashed and rebased.

File: src/IO/Observer/ReductionActions.hpp (hunk 2 above)
Comment (nilsdeppe) [reply src/IO/Observer/ReductionActions.hpp]: ping on `noexcept`

File: src/IO/Observer/ReductionActions.hpp (hunk 3 above)
Comment (nilsdeppe) [reply src/IO/Observer/ReductionActions.hpp]: ping on `noexcept`

Comment (carmaza) [reply src/IO/Observer/ReductionActions.hpp]: done

**Review by nilsdeppe — APPROVED** [review]
LGTM! Thanks for implementing this!

image

**Review by nilsvu — APPROVED** [review]
LGTM 👍 I'm not entirely sure this is necessary for #2048 but having the functionality won't hurt, so we can merge this and I'll look at #2048 next.

**Pull request merged**